### PR TITLE
HPCC-16345 Optimize MAX(unsigned, 0) and IF(unsigned < 0, 0, unsigned)

### PR DESCRIPTION
**Pull request by richardkchapman**
Helps avoid C++ compiler warnings.

Signed-off-by: Richard Chapman rchapman@hpccsystems.com

### CODE REVIEW

**Comment by richardkchapman** [comment]
@ghalliday Please review

File: ecl/hql/hqlfold.cpp [hunk 1]
```diff
@@ -450,6 +450,17 @@ static IHqlExpression * optimizeCompare(IHqlExpression * expr)
...
     {
         ITypeInfo * rType = rightChild->queryType();
+        bool isUnsigned = rType->getTypeCode()==type_int && !rType->isSigned();
+        if (isUnsigned && leftValue->getIntValue()==0)
+        {
+            switch (op)
+            {
+            case no_le:
+                return createConstant(true);
+            case no_ge:
```
Comment (ghalliday) [inline ecl/hql/hqlfold.cpp]: Should be no_gt

File: ecl/hql/hqlfold.cpp [hunk 2]
```diff
@@ -481,6 +492,17 @@ static IHqlExpression * optimizeCompare(IHqlExpression * expr)
     else if (rightValue)
     {
         ITypeInfo * lType = leftChild->queryType();
+        bool isUnsigned = lType->getTypeCode()==type_int && !lType->isSigned();
+        if (isUnsigned && rightValue->getIntValue()==0)
```
Comment (ghalliday) [inline ecl/hql/hqlfold.cpp]: could use isZero(rightChild)

File: ecl/hql/hqlfold.cpp [hunk 3]
```diff
@@ -481,6 +492,17 @@ static IHqlExpression * optimizeCompare(IHqlExpression * expr)
     else if (rightValue)
     {
         ITypeInfo * lType = leftChild->queryType();
+        bool isUnsigned = lType->getTypeCode()==type_int && !lType->isSigned();
```
Comment (ghalliday) [inline ecl/hql/hqlfold.cpp]: Could also match decimal, swapint, and packed int.  Worth extracting into an isUnsignedType()

File: ecl/hql/hqlfold.cpp [hunk 4]
```diff
@@ -3913,6 +3936,11 @@ IHqlExpression * foldConstantOperator(IHqlExpression * expr, unsigned foldOption
                         IValue * value = cur->queryValue();
                         if (value)
                         {
+                            if (isUnsigned && value->getIntValue()==0 && op==no_maxlist)
+                            {
+                                same = false;
+                                continue;
+                            }
```
Comment (ghalliday) [inline ecl/hql/hqlfold.cpp]: unlikely to catch anything, but if no_minlist it could return LINK(cur).

**Review by ghalliday — CHANGES_REQUESTED** [review]
@richardkchapman this also needs some test cases.

**Comment by richardkchapman** [comment]
Repushed to address comments - @ghalliday Please re-review

**Comment by ghalliday** [comment]
@richardkchapman I'll run the regression test.  Curious why you didn't also support decimal and swap int.

**Comment by ghalliday** [comment]
It still needs a test case or two...

**Review by ghalliday — APPROVED** [review]

**Pull request merged**